// react-vant/lib/empty/Empty.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
function _react() {
  const data = _interopRequireWildcard(require("react"));
  _react = function () {
    return data;
  };
  return data;
}
function _jsxRuntime() {
  const data = require("react/jsx-runtime");
  _jsxRuntime = function () {
    return data;
  };
  return data;
}
function _clsx() {
  const data = _interopRequireDefault(require("clsx"));
  _clsx = function () {
    return data;
  };
  return data;
}
var _utils = require("../utils");
var _Network = require("./Network");
var _getDefaultProps = require("../utils/get-default-props");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && Object.prototype.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
const PRESET_IMAGES = ['error', 'search', 'default'];
const [bem] = (0, _utils.createNamespace)('empty');
const Empty = p => {
  const props = (0, _getDefaultProps.mergeProps)(p, {
    image: 'default'
  });
  const renderImage = () => {
    let {
      image
    } = props;
    if ((0, _react().isValidElement)(image)) {
      return image;
    }
    if (image === 'network') {
      return _Network.Network;
    }
    if (PRESET_IMAGES.includes(image)) {
      image = `https://img.yzcdn.cn/vant/empty-image-${image}.png`;
    }
    return (0, _jsxRuntime().jsx)("img", {
      src: image,
      alt: ''
    });
  };
  const renderDescription = () => {
    if (props.description) {
      return (0, _jsxRuntime().jsx)("div", Object.assign({
        className: (0, _clsx().default)(bem('description'))
      }, {
        children: props.description
      }));
    }
    return null;
  };
  const renderBottom = () => {
    if (props.children) {
      return (0, _jsxRuntime().jsx)("div", Object.assign({
        className: (0, _clsx().default)(bem('bottom'))
      }, {
        children: props.children
      }));
    }
    return null;
  };
  return (0, _jsxRuntime().jsxs)("div", Object.assign({
    className: (0, _clsx().default)(props.className, bem()),
    style: props.style
  }, {
    children: [(0, _jsxRuntime().jsx)("div", Object.assign({
      className: (0, _clsx().default)(bem('image')),
      style: (0, _utils.getSizeStyle)(props.imageSize)
    }, {
      children: renderImage()
    })), renderDescription(), renderBottom()]
  }));
};
var _default = exports.default = Empty;
