// react-vant/lib/area/Area.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
function _react() {
  const data = _interopRequireWildcard(require("react"));
  _react = function () {
    return data;
  };
  return data;
}
function _jsxRuntime() {
  const data = require("react/jsx-runtime");
  _jsxRuntime = function () {
    return data;
  };
  return data;
}
function _clsx() {
  const data = _interopRequireDefault(require("clsx"));
  _clsx = function () {
    return data;
  };
  return data;
}
var _picker = _interopRequireDefault(require("../picker"));
var _utils = require("../utils");
var _getDefaultProps = require("../utils/get-default-props");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && Object.prototype.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
const INHERIT_PROPS = ['title', 'visible', 'popup', 'value', 'defaultValue', 'loading', 'readOnly', 'itemHeight', 'swipeDuration', 'visibleItemCount', 'cancelButtonText', 'confirmButtonText', 'toolbar', 'columnsTop', 'columnsBottom', 'optionRender', 'placeholder', 'onChange', 'onCancel', 'onConfirm', 'children'];
function parseVanAreaList(data, columnsNum) {
  const {
    province_list = {},
    city_list = {},
    county_list = {}
  } = data;
  const provinces = Object.entries(province_list).map(([value, text]) => ({
    value,
    text
  }));
  const citys = Object.entries(city_list).map(([value, text]) => ({
    value,
    text
  }));
  const countrys = Object.entries(county_list).map(([value, text]) => ({
    value,
    text
  }));
  if (columnsNum > 2) {
    citys.forEach(city => {
      var _a;
      const value = (_a = city.value) === null || _a === void 0 ? void 0 : _a.slice(0, 4);
      const children = countrys.filter(country => {
        var _a;
        return ((_a = country.value) === null || _a === void 0 ? void 0 : _a.slice(0, 4)) === value;
      });
      city.children = children;
    });
  }
  if (columnsNum > 1) {
    provinces.forEach(province => {
      var _a;
      const provinceCode = (_a = province.value) === null || _a === void 0 ? void 0 : _a.slice(0, 2);
      const children = citys.filter(city => {
        var _a;
        return ((_a = city.value) === null || _a === void 0 ? void 0 : _a.slice(0, 2)) === provinceCode;
      });
      province.children = children;
    });
  }
  return provinces;
}
const [bem] = (0, _utils.createNamespace)('area');
const Area = (0, _react().forwardRef)((p, ref) => {
  const props = (0, _getDefaultProps.mergeProps)(p, {
    areaList: {},
    columnsNum: 3,
    itemHeight: 44,
    visibleItemCount: 5,
    swipeDuration: 300,
    showToolbar: true,
    placeholder: true,
    toolbarPosition: 'top'
  });
  const columns = (0, _react().useMemo)(() => {
    var _a;
    return (_a = props.columns) !== null && _a !== void 0 ? _a : parseVanAreaList(props.areaList, +props.columnsNum);
  }, [props.columnsNum, props.areaList, props.columns]);
  return (0, _jsxRuntime().jsx)(_picker.default, Object.assign({
    ref: ref,
    style: props.style,
    className: (0, _clsx().default)(bem(), props.className),
    columns: columns
  }, (0, _utils.pick)(props, INHERIT_PROPS)));
});
var _default = exports.default = Area;
