// react-vant/lib/sku/components/SkuRowPropItem.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
function _react() {
  const data = _interopRequireWildcard(require("react"));
  _react = function () {
    return data;
  };
  return data;
}
function _jsxRuntime() {
  const data = require("react/jsx-runtime");
  _jsxRuntime = function () {
    return data;
  };
  return data;
}
function _clsx() {
  const data = _interopRequireDefault(require("clsx"));
  _clsx = function () {
    return data;
  };
  return data;
}
var _utils = require("../../utils");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && Object.prototype.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
const [bem] = (0, _utils.createNamespace)('sku-row');
const SkuRowPropItem = props => {
  var _a;
  const choosed = (0, _react().useMemo)(() => {
    const {
      selectedProp,
      skuKeyStr,
      skuValue
    } = props;
    if (selectedProp && selectedProp[skuKeyStr]) {
      return selectedProp[skuKeyStr].indexOf(skuValue.id) > -1;
    }
    return false;
  }, [JSON.stringify(props.selectedProp), JSON.stringify(props.skuValue), props.skuKeyStr]);
  const onSelect = () => {
    props.onSkuPropSelected(Object.assign(Object.assign({}, props.skuValue), {
      skuKeyStr: props.skuKeyStr,
      multiple: props.multiple
    }));
  };
  return (0, _jsxRuntime().jsx)("span", Object.assign({
    className: (0, _clsx().default)(bem('item', {
      active: choosed
    })),
    onClick: onSelect
  }, {
    children: (0, _jsxRuntime().jsx)("span", Object.assign({
      className: (0, _clsx().default)(bem('item-name'))
    }, {
      children: (_a = props.skuValue) === null || _a === void 0 ? void 0 : _a.name
    }))
  }));
};
var _default = exports.default = SkuRowPropItem;
