// react-vant/lib/sku/components/SkuStepper.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
function _react() {
  const data = _interopRequireWildcard(require("react"));
  _react = function () {
    return data;
  };
  return data;
}
function _jsxRuntime() {
  const data = require("react/jsx-runtime");
  _jsxRuntime = function () {
    return data;
  };
  return data;
}
function _clsx() {
  const data = _interopRequireDefault(require("clsx"));
  _clsx = function () {
    return data;
  };
  return data;
}
var _constants = require("../constants");
var _stepper = _interopRequireDefault(require("../../stepper"));
var _utils = require("../../utils");
var _getDefaultProps = require("../../utils/get-default-props");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && Object.prototype.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
const {
  QUOTA_LIMIT,
  STOCK_LIMIT
} = _constants.LIMIT_TYPE;
const [bem] = (0, _utils.createNamespace)('sku');
const SkuStepper = p => {
  const props = (0, _getDefaultProps.mergeProps)(p, {
    quota: 0,
    quotaUsed: 0,
    startSaleNum: 1,
    customStepperConfig: {}
  });
  const [limitType, setLimitType] = (0, _react().useState)(STOCK_LIMIT);
  const stepperLimit = (0, _react().useMemo)(() => {
    const quotaLimit = +props.quota - +props.quotaUsed;
    let limit;
    let currentLimitType;
    // 无限购时直接取库存，有限购时取限购数和库存数中小的那个
    if (+props.quota > 0 && quotaLimit <= props.stock) {
      // 修正负的limit
      limit = quotaLimit < 0 ? 0 : quotaLimit;
      currentLimitType = QUOTA_LIMIT;
    } else {
      limit = props.stock;
      currentLimitType = STOCK_LIMIT;
    }
    setLimitType(currentLimitType);
    return limit;
  }, [props.quota, props.quotaUsed, props.stock]);
  const stepperMinLimit = (0, _react().useMemo)(() => {
    return +props.startSaleNum < 1 ? 1 : +props.startSaleNum;
  }, [props.startSaleNum]);
  const quotaContent = (0, _react().useMemo)(() => {
    const {
      quotaText,
      hideQuotaText
    } = props.customStepperConfig;
    if (hideQuotaText) return '';
    let text = '';
    if (quotaText) {
      text = quotaText;
    } else {
      const textArr = [];
      if (+props.startSaleNum > 1) {
        textArr.push(`${props.startSaleNum}件起售`);
      }
      if (+props.quota > 0) {
        textArr.push(`限购${+props.quota}件`);
      }
      text = textArr.join(',');
    }
    return text;
  }, [JSON.stringify(props.customStepperConfig), props.startSaleNum, props.quota]);
  const checkState = (min, max) => {
    // 如果选择小于起售，则强制变为起售
    if (props.currentNum < min || min > max) {
      props.onChange(min);
    } else if (props.currentNum > max) {
      // 当前选择数量大于最大可选时，需要重置已选数量
      props.onChange(max);
    }
    props.onSkuStepperState({
      valid: min <= max,
      min,
      max,
      limitType,
      quota: props.quota,
      quotaUsed: props.quotaUsed,
      startSaleNum: props.startSaleNum
    });
  };
  const setCurrentNum = num => {
    props.onChange(num);
    checkState(stepperMinLimit, stepperLimit);
  };
  const onChange = currentValue => {
    const intValue = parseInt(currentValue, 10);
    const {
      handleStepperChange
    } = props.customStepperConfig;
    if (handleStepperChange) handleStepperChange(intValue);
    setCurrentNum(currentValue);
  };
  const onOverLimit = action => {
    props.onSkuOverLimit({
      action,
      limitType,
      quota: props.quota,
      quotaUsed: props.quotaUsed,
      startSaleNum: props.startSaleNum
    });
  };
  return (0, _jsxRuntime().jsxs)("div", Object.assign({
    className: (0, _clsx().default)(bem('stepper-stock'))
  }, {
    children: [(0, _jsxRuntime().jsx)("div", Object.assign({
      className: (0, _clsx().default)(bem('stepper-title'))
    }, {
      children: props.stepperTitle
    })), (0, _jsxRuntime().jsx)(_stepper.default, {
      min: stepperMinLimit,
      max: stepperLimit,
      className: (0, _clsx().default)(bem('stepper')),
      disableInput: props.disableStepperInput,
      value: props.currentNum,
      onChange: onChange,
      onOverlimit: onOverLimit
    }), !props.hideQuotaText && quotaContent && (0, _jsxRuntime().jsx)("span", Object.assign({
      className: (0, _clsx().default)(bem('stepper-quota'))
    }, {
      children: quotaContent
    }))]
  }));
};
var _default = exports.default = SkuStepper;
