// react-vant/lib/text-area/TextArea.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
function _jsxRuntime() {
  const data = require("react/jsx-runtime");
  _jsxRuntime = function () {
    return data;
  };
  return data;
}
function _react() {
  const data = _interopRequireWildcard(require("react"));
  _react = function () {
    return data;
  };
  return data;
}
function _icons() {
  const data = require("@react-vant/icons");
  _icons = function () {
    return data;
  };
  return data;
}
function _clsx() {
  const data = _interopRequireDefault(require("clsx"));
  _clsx = function () {
    return data;
  };
  return data;
}
var _utils = require("../utils");
var _hooks = require("../hooks");
var _getDefaultProps = require("../utils/get-default-props");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && Object.prototype.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
const [bem] = (0, _utils.createNamespace)('textarea');
const TextArea = (0, _react().forwardRef)((p, ref) => {
  const props = (0, _getDefaultProps.mergeProps)(p, {
    rows: 2,
    clearIcon: (0, _jsxRuntime().jsx)(_icons().Clear, {}),
    clearTrigger: 'focus',
    defaultValue: ''
  });
  const [hasFocus, setHasFocus] = (0, _react().useState)(false);
  const nativeTextAreaRef = (0, _react().useRef)();
  const compositionStartRef = (0, _react().useRef)(false);
  const [value, setValue] = (0, _hooks.usePropsValue)(props);
  const {
    className,
    style,
    name,
    rows,
    placeholder,
    disabled,
    readOnly,
    maxLength,
    showWordLimit,
    autoFocus
  } = props;
  const focus = () => {
    if (nativeTextAreaRef === null || nativeTextAreaRef === void 0 ? void 0 : nativeTextAreaRef.current) {
      nativeTextAreaRef.current.focus();
    }
  };
  const blur = () => {
    if (nativeTextAreaRef === null || nativeTextAreaRef === void 0 ? void 0 : nativeTextAreaRef.current) {
      nativeTextAreaRef.current.blur();
    }
  };
  (0, _react().useImperativeHandle)(ref, () => ({
    clear: () => {
      setValue('');
    },
    focus,
    blur,
    get nativeElement() {
      return nativeTextAreaRef.current;
    }
  }));
  const adjustSize = () => {
    const input = nativeTextAreaRef.current;
    if (!input) return;
    input.style.height = 'auto';
    let height = input.scrollHeight;
    if ((0, _utils.isObject)(props.autoSize)) {
      const {
        maxHeight,
        minHeight
      } = props.autoSize;
      if (maxHeight) {
        height = Math.min(height, maxHeight);
      }
      if (minHeight) {
        height = Math.max(height, minHeight);
      }
    }
    if (height) {
      input.style.height = `${height}px`;
    }
  };
  (0, _react().useEffect)(() => {
    adjustSize();
  }, [value]);
  const controlClass = _react().default.useMemo(() => {
    return bem('control', [{
      'min-height': !props.autoSize,
      clear: props.clearable
    }]);
  }, [props.autoSize]);
  const handleValueChange = inputValue => {
    var _a;
    let finalValue = inputValue;
    if (!compositionStartRef.current) {
      if ((0, _utils.isDef)(maxLength) && finalValue.length > +maxLength) {
        finalValue = finalValue.slice(0, maxLength);
        (_a = props.onOverlimit) === null || _a === void 0 ? void 0 : _a.call(props);
      }
    }
    setValue(finalValue);
  };
  const handleFocus = e => {
    var _a;
    setHasFocus(true);
    (_a = props.onFocus) === null || _a === void 0 ? void 0 : _a.call(props, e);
    // readOnly not work in legacy mobile safari
    if (readOnly) {
      blur();
    }
  };
  const handleBulr = e => {
    var _a;
    setHasFocus(false);
    (_a = props.onBlur) === null || _a === void 0 ? void 0 : _a.call(props, e);
    (0, _utils.resetScroll)();
  };
  const renderWordLimit = () => {
    if (showWordLimit) {
      const currentCount = (value ? `${value}` : '').length;
      return (0, _jsxRuntime().jsx)("div", Object.assign({
        className: (0, _clsx().default)(bem('word-limit'))
      }, {
        children: typeof showWordLimit === 'function' ? showWordLimit({
          currentCount,
          maxLength
        }) : (0, _jsxRuntime().jsxs)(_jsxRuntime().Fragment, {
          children: [(0, _jsxRuntime().jsx)("span", Object.assign({
            className: (0, _clsx().default)(bem('word-num'))
          }, {
            children: currentCount
          })), maxLength ? `/${maxLength}` : false]
        })
      }));
    }
    return null;
  };
  const handleClear = e => {
    var _a;
    setValue('');
    (_a = props.onClear) === null || _a === void 0 ? void 0 : _a.call(props, e);
  };
  const showClear = (0, _react().useMemo)(() => {
    if (props.clearable && !readOnly) {
      const hasValue = value !== '';
      const trigger = props.clearTrigger === 'always' || props.clearTrigger === 'focus' && hasFocus;
      return hasValue && trigger;
    }
    return false;
  }, [value, props.clearTrigger, hasFocus]);
  return (0, _jsxRuntime().jsxs)("div", Object.assign({
    className: (0, _clsx().default)(bem(), className),
    style: style
  }, {
    children: [(0, _jsxRuntime().jsx)("textarea", {
      ref: nativeTextAreaRef,
      name: name,
      rows: rows,
      className: (0, _clsx().default)(controlClass),
      value: value,
      disabled: disabled,
      autoFocus: autoFocus,
      readOnly: readOnly,
      placeholder: placeholder || '',
      onBlur: handleBulr,
      onFocus: handleFocus,
      onChange: e => {
        var _a;
        return handleValueChange((_a = e === null || e === void 0 ? void 0 : e.currentTarget) === null || _a === void 0 ? void 0 : _a.value);
      },
      onKeyPress: props.onKeyPress,
      onKeyDown: props.onKeyDown,
      onKeyUp: props.onKeyUp,
      autoComplete: props.autoComplete,
      onCompositionStart: e => {
        var _a;
        compositionStartRef.current = true;
        (_a = props.onCompositionStart) === null || _a === void 0 ? void 0 : _a.call(props, e);
      },
      onCompositionEnd: e => {
        var _a, _b;
        compositionStartRef.current = false;
        (_a = props.onCompositionEnd) === null || _a === void 0 ? void 0 : _a.call(props, e);
        handleValueChange((_b = e === null || e === void 0 ? void 0 : e.currentTarget) === null || _b === void 0 ? void 0 : _b.value);
      },
      onClick: props.onClick
    }), showClear && _react().default.cloneElement(props.clearIcon, {
      className: (0, _clsx().default)(bem('clear')),
      onTouchStart: handleClear
    }), renderWordLimit()]
  }));
});
var _default = exports.default = TextArea;
