// react-vant/lib/utils/constant.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.WHITE = exports.SHADOW = exports.RED = exports.GREEN = exports.GRAY_DARK = exports.GRAY = exports.COMPONENT_TYPE_KEY = exports.BORDER_UNSET_TOP_BOTTOM = exports.BORDER_TOP_BOTTOM = exports.BORDER_TOP = exports.BORDER_SURROUND = exports.BORDER_RIGHT = exports.BORDER_LEFT = exports.BORDER_BOTTOM = exports.BORDER = exports.BLUE = void 0;
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
// color
const RED = exports.RED = '#ff603f';
const BLUE = exports.BLUE = '#2879ff';
const GREEN = exports.GREEN = '#00c853';
const WHITE = exports.WHITE = '#fff';
const GRAY = exports.GRAY = '#c9c9c9';
const GRAY_DARK = exports.GRAY_DARK = '#969799';
// border
const BORDER = exports.BORDER = 'rv-hairline';
const BORDER_TOP = exports.BORDER_TOP = `${BORDER}--top`;
const BORDER_LEFT = exports.BORDER_LEFT = `${BORDER}--left`;
const BORDER_RIGHT = exports.BORDER_RIGHT = `${BORDER}--right`;
const BORDER_BOTTOM = exports.BORDER_BOTTOM = `${BORDER}--bottom`;
const BORDER_SURROUND = exports.BORDER_SURROUND = `${BORDER}--surround`;
const BORDER_TOP_BOTTOM = exports.BORDER_TOP_BOTTOM = `${BORDER}--top-bottom`;
const BORDER_UNSET_TOP_BOTTOM = exports.BORDER_UNSET_TOP_BOTTOM = `${BORDER}-unset--top-bottom`;
// shadow
const SHADOW = exports.SHADOW = 'rv-shadow';
// component type key
const COMPONENT_TYPE_KEY = exports.COMPONENT_TYPE_KEY = '__REACT_VANT_COMPONENT';
